File: scripts/fix_tax_issued.js
// 세금계산서 발행여부 복구: CSV에서 발행완료인 건을 Supabase에 업데이트
const fs = require('fs')
const path = require('path')

const env = fs.readFileSync(path.join(__dirname, '../.env.local'), 'utf8')
const url = env.match(/NEXT_PUBLIC_SUPABASE_URL=(.+)/)?.[1]?.trim()
const key = env.match(/SUPABASE_SERVICE_ROLE_KEY=(.+)/)?.[1]?.trim()
const headers = {
  apikey: key,
  Authorization: 'Bearer ' + key,
  'Content-Type': 'application/json',
  Prefer: 'return=minimal',
}

async function fetchAll(table, params) {
  const r = await fetch(`${url}/rest/v1/${table}?${params}`, { headers })
  return r.json()
}

async function updateById(table, id, body) {
  const r = await fetch(`${url}/rest/v1/${table}?id=eq.${id}`, {
    method: 'PATCH',
    headers,
    body: JSON.stringify(body),
  })
  return r.status
}

;(async () => {
  // CSV 읽기
  const csvFiles = fs.readdirSync('C:\\Users\\Win11\\Downloads\\gradius_python-master (1)\\gradius_python-master\\migration_csvs')
  // 청구수납금액입력 파일 찾기 (크기로 구분 - 가장 큰 파일)
  const csvDir = 'C:\\Users\\Win11\\Downloads\\gradius_python-master (1)\\gradius_python-master\\migration_csvs'
  const files = fs.readdirSync(csvDir).map(f => ({
    name: f,
    size: fs.statSync(path.join(csvDir, f)).size,
    fullPath: path.join(csvDir, f),
  }))
  const biggestFile = files.sort((a, b) => b.size - a.size)[0]
  console.log('읽는 파일:', biggestFile.name, biggestFile.size, 'bytes')

  const raw = fs.readFileSync(biggestFile.fullPath, 'utf8')
  const lines = raw.split('\n').filter(l => l.trim())
  
  // 헤더 파싱 (간단히 쉼표 분리)
  const headerLine = lines[0]
  const headers2 = headerLine.split(',').map(h => h.replace(/"/g, '').trim())
  console.log('컬럼 수:', headers2.length)
  console.log('컬럼:', headers2.join(' | '))

  // 인덱스 찾기
  const idIdx = 0  // 청구ID는 첫 번째
  const taxIdx = headers2.findIndex(h => h.includes('발행여부'))
  console.log('발행여부 컬럼 idx:', taxIdx, ':', headers2[taxIdx])

  if (taxIdx === -1) {
    console.log('ERROR: 발행여부 컬럼을 찾을 수 없습니다')
    return
  }

  // 발행완료 ID 수집
  const issuedCsvIds = []
  for (let i = 1; i < Math.min(lines.length, 200); i++) {
    // 단순 쉼표 분리 (따옴표 안의 쉼표 무시 - 간단 처리)
    const cols = lines[i].split(',')
    if (cols.length <= taxIdx) continue
    
    const csvId = cols[idIdx].replace(/"/g, '').trim()
    const taxVal = cols[taxIdx].replace(/"/g, '').trim()
    
    if (taxVal === '발행완료') {
      issuedCsvIds.push(csvId)
    }
  }
  
  console.log(`\n발행완료 건수: ${issuedCsvIds.length}개`)
  console.log('샘플 ID:', issuedCsvIds.slice(0, 5))

  // Supabase settlements에서 해당 건 찾기
  // settlements 테이블에 inquiry_id로 연결되어 있으므로
  // inquiries 테이블에서 old_id (청구ID에 해당하는 값)로 매칭
  const settlements = await fetchAll('settlements', 'select=id,company_name,inquiry_id,site_name,tax_invoice_issued&limit=200')
  console.log('\nSettlements 총:', settlements.length, '건')
  
  // 방법: progress = '정산완료' + deposit_status = '입금완료' 인 건을 발행완료로 처리
  // CSV ID 매칭이 어려우므로, CSV의 발행완료 비율을 확인 후
  // Supabase에서 정산완료 건들을 업데이트
  
  // 먼저 CSV 전체 파싱해서 발행완료 비율 확인
  let totalCsv = 0
  const csvData = {}
  for (let i = 1; i < lines.length; i++) {
    const cols = lines[i].split(',')
    if (cols.length <= taxIdx) continue
    totalCsv++
    const csvId = cols[idIdx].replace(/"/g, '').trim()
    const taxVal = cols[taxIdx].replace(/"/g, '').trim()
    csvData[csvId] = taxVal
  }
  
  const issuedCount = Object.values(csvData).filter(v => v === '발행완료').length
  const notIssuedCount = Object.values(csvData).filter(v => v !== '발행완료').length
  console.log(`\nCSV 분포: 발행완료=${issuedCount} / 미발행 등=${notIssuedCount} / 전체=${totalCsv}`)

  // progress = 정산완료인 settlements를 발행완료로 업데이트
  const toUpdate = settlements.filter(s => {
    // CSV 발행완료 비율이 높다면 정산완료 = 발행완료로 봄
    return false // 일단 확인만 - 실제 업데이트는 사용자 확인 후
  })
  
  console.log('\n=== 최종 확인 ===')
  console.log('CSV 발행완료:', issuedCount, '건')
  console.log('Supabase 정산완료:', settlements.filter(s => s.progress === '정산완료').length, '건')
  console.log('Supabase 행사종료+입금완료:', settlements.filter(s => s.progress === '행사종료' && s.deposit_status === '입금완료').length, '건')
})()

File: scripts/fix_tax_issued.py
#!/usr/bin/env python3
# 세금계산서 발행여부 복구 스크립트
# CSV에서 '발행완료'인 건을 찾아 Supabase settlements 테이블 업데이트

import csv
import os
import json
import urllib.request
import urllib.parse

# 환경변수 읽기
env_path = os.path.join(os.path.dirname(__file__), '../.env.local')
env = {}
with open(env_path, encoding='utf-8') as f:
    for line in f:
        line = line.strip()
        if '=' in line and not line.startswith('#'):
            k, v = line.split('=', 1)
            env[k.strip()] = v.strip()

SUPABASE_URL = env.get('NEXT_PUBLIC_SUPABASE_URL', '')
SUPABASE_KEY = env.get('SUPABASE_SERVICE_ROLE_KEY', '')

HEADERS = {
    'apikey': SUPABASE_KEY,
    'Authorization': f'Bearer {SUPABASE_KEY}',
    'Content-Type': 'application/json',
    'Prefer': 'return=minimal',
}

def supabase_get(table, params=''):
    url = f"{SUPABASE_URL}/rest/v1/{table}?{params}"
    req = urllib.request.Request(url, headers=HEADERS)
    with urllib.request.urlopen(req) as r:
        return json.loads(r.read())

def supabase_patch(table, filter_param, body):
    url = f"{SUPABASE_URL}/rest/v1/{table}?{filter_param}"
    data = json.dumps(body).encode('utf-8')
    req = urllib.request.Request(url, data=data, headers=HEADERS, method='PATCH')
    with urllib.request.urlopen(req) as r:
        return r.status

# CSV 파일 경로
csv_dir = r'C:\Users\Win11\Downloads\gradius_python-master (1)\gradius_python-master\migration_csvs'
# 가장 큰 파일 찾기
files = [(f, os.path.getsize(os.path.join(csv_dir, f))) for f in os.listdir(csv_dir)]
files.sort(key=lambda x: x[1], reverse=True)
csv_file = os.path.join(csv_dir, files[0][0])
print(f"CSV 파일: {files[0][0]} ({files[0][1]:,} bytes)")

# CSV 파싱 (python csv 모듈로 올바르게 파싱)
issued_ids = []  # 발행완료 문의ID
all_data = {}    # {문의ID: {진행상황, 발행여부, ...}}

with open(csv_file, encoding='utf-8-sig', newline='') as f:
    reader = csv.DictReader(f)
    for row in reader:
        inquiry_id = row.get('문의ID', '').strip()
        tax_status = row.get('세금계산서 발행여부', '').strip()
        progress = row.get('진행상황', '').strip()
        deposit = row.get('입금여부', '').strip()
        
        all_data[inquiry_id] = {
            'tax_status': tax_status,
            'progress': progress,
            'deposit': deposit,
            'company': row.get('업체', '').strip(),
        }
        
        if tax_status == '발행완료':
            issued_ids.append(inquiry_id)

print(f"\n전체 CSV 건수: {len(all_data)}")
print(f"발행완료: {len(issued_ids)}건")

# 발행여부 값 분포
from collections import Counter
tax_vals = Counter(v['tax_status'] for v in all_data.values())
print(f"\n세금계산서 발행여부 값 분포: {dict(tax_vals)}")

progress_vals = Counter(v['progress'] for v in all_data.values())
print(f"진행상황 값 분포: {dict(progress_vals)}")

# Supabase settlements 조회
print("\n\n=== Supabase settlements 현황 ===")
settlements = supabase_get('settlements', 'select=id,inquiry_id,company_name,tax_invoice_issued,progress,deposit_status&limit=200')
print(f"총 {len(settlements)}건")

sett_by_inq = {str(s['inquiry_id']): s for s in settlements if s.get('inquiry_id')}
print(f"inquiry_id 있는 건: {len(sett_by_inq)}")

# 발행완료인 CSV ID와 settlements 매칭
matched_issued = []
for csvid in issued_ids:
    if csvid in sett_by_inq:
        matched_issued.append(sett_by_inq[csvid])

print(f"\nCSV 발행완료 {len(issued_ids)}건 중 settlements 매칭: {len(matched_issued)}건")
if matched_issued:
    print("샘플:")
    for s in matched_issued[:5]:
        print(f"  inquiry_id={s['inquiry_id']}, company={s['company_name']}, 현재 tax_invoice_issued={s['tax_invoice_issued']}")

# 실제 업데이트
print(f"\n\n=== Supabase 업데이트 시작 ===")
updated = 0
failed = 0

for sett in matched_issued:
    if sett['tax_invoice_issued'] == True:
        print(f"  이미 True: {sett['company_name']}")
        continue
    try:
        status = supabase_patch('settlements', f"id=eq.{sett['id']}", {'tax_invoice_issued': True})
        if status in (200, 204):
            updated += 1
            print(f"  [OK] {sett['company_name']} -> tax_invoice_issued=true")
        else:
            failed += 1
            print(f"  [FAIL] {sett['company_name']}, status={status}")
    except Exception as e:
        failed += 1
        print(f"  [ERROR] {sett['company_name']}: {e}")

print(f"\n완료: 업데이트={updated}, 실패={failed}")
print(f"(CSV 발행완료 {len(issued_ids)}건 중 매칭={len(matched_issued)}건 업데이트 시도)")
